### ISSUE TITLE
Different behaviour in np.array between object/scalar 0d arrays in iterable

### ISSUE THREAD

**Issue by Gattocrucco** [numpy/numpy#16052]
```python
import numpy as np
x = object()
np.array([array(1)]) # array([1])
np.array(array(x)) # array(<object object at 0x11c4415d0>, dtype=object)
np.array([x]) # array([<object object at 0x11c4415d0>], dtype=object)
np.array([array(x)]) # array([array(<object object at 0x11c4415d0>, dtype=object)], dtype=object)
```

The last result is inconsistent. It should be `array([<object object at 0x11c4415d0>], dtype=object)`.

### Numpy/Python version information:

```
1.18.2 3.8.2 (v3.8.2:7b3ab5921f, Feb 24 2020, 17:52:18) 
[Clang 6.0 (clang-600.0.57)]
```

**Comment by eric-wieser** [comment]
Thanks for the report. I don't think cases 2 and 3 are particularly interesting, but I can see the argument that 1 and 4 are inconsistent.

There's another way to show the inconsistency of 4:
```python
# nested 2d arrays - unpacked
>>> np.array([np.array([[None]]), np.array([[None]])])
array([[[None]],

       [[None]]], dtype=object)

# nested 1d arrays - unpacked
>>> np.array([np.array([None]), np.array([None])])
array([[None],
       [None]], dtype=object)

# nested 0d arrays - not unpacked!
>>> np.array([np.array(None), np.array(None)])
 array([array(None, dtype=object), array(None, dtype=object)], dtype=object)
```

(Note: `None` is an easy way to get an object array)

**Comment by vrakesh** [comment]
@eric-wieser I want to help on this one, I looked at the code, for array creation, I am still not certain how the unpacking happens for 1D array, could you point me in the direction of the code for this?

**Comment by seberg** [comment]
The actual code that does this is in `setArrayFromSequence`, which calls `PyArray_SETITEM` at some point. So if you wanted to tweak it, that `PyArray_SETITEM` is where you have to do that.

**Comment by vrakesh** [comment]
@seberg , sorry for the late response on this, got side tracked, after looking into this.

Here is what I understand, 

We start with the lowest dimension and copy over as we go in ```setArrayFromSequence``` 

so for instance

When invoking 

```python
np.array([None])

# for a sequence length of 1 it looks at NoneType and copies over it as intended

np.array(np.array([None])

# Also works as intended, the lowest NoneType is detected correctly and rest is just a copy call

# but when calling

np.array(np.array(None))

# The lowest dim seen is np.array not NoneType
```

Hence the solution is to do a ```(PyArrayObject *)array_item_asarray(dst, i);``` not just when

```(PyArray_NDIM(a) - dim) > 1 ```

but when the lowest dimension in the sequence is an np.array

i.e 

replace lines 560 and 590 in ctors.c

with
```c
 if ((PyArray_NDIM(a) - dim) > 1 || strcmp(Py_TYPE(o)->tp_name,"numpy.ndarray") == 0) {
                PyArrayObject * tmp =
                    (PyArrayObject *)array_item_asarray(dst, i);
```

Am I in the right direction?? Is there a neater way to do this check?

**Comment by seberg** [comment]
Yeah, that is basically correct. To say that you need to enter the scalar dimension if this is an array... It gets more complicated, and I think impossible to fix without a larger rework (and I am trying to do that right now).
Reading the code, this should also fail for things like this:
```
import numpy as np

class arraylike:
    def __array__(self, dtype=None):
        return np.array(())  # 0-D array, yay!

np.array(arraylike())  # OK
np.array([1, arraylike()])  # boom
```
Frankly, it may be better to not try to patch this one up, until we patch this hole thing up for "good". Not to say I do not encourage you to look at it more :)!

I am currently working on a 2-pass approach, which caches arrays and sequences found on the first path, which would make the necessary information much nicer available here.

**Comment by seberg** [comment]
Sorry, I suppose the thing is that if this the other array is 0-D, then it should be unpacked. But if it is not 0-D, then SETITEM should be used. That does not really help much with my above example, but can be made to work for object case I guess...

**Comment by seberg** [comment]
Honestly, I am just trying to fix this up (on my branch, not in master). And it is super ugly...
Consider all of these cases:
```
arr = np.array(0)
arr_ndim = np.array([0])

np.array(arr)  # unpack
np.array(arr_ndim)  # unpack

np.array([arr, arr], dtype=object)  # unpack
np.array([arr, None], dtype=object)  # unpack!?
np.array([arr_ndim, None], dtype=object)  # cannot unpack

np.array([arr_ndim, arr])  # ragged, but similar to the previous
```
Yes, you *can* distinguish `np.array([None, arr])`, from `np.array([arr, arr])` but it seems a bit crazy. To be honest, I think I like the current behaviour best right now. Especially, once you factor in array likes, which we really do not want to unpack probably...

The good rule would be "unpack exactly one level deep", but we do not have a well defined level in the mixed cases, and especially the ragged cases really would have to just fail in that case.

**Comment by vrakesh** [comment]
I agree, this is indeed pretty tricky, the more I look into it, too many corner cases, it is hard to satisfy hard, without complete rework

**Comment by seberg** [comment]
I have a complete rework in an alpha stage hopefully (tests pass), needs some finishing up though. And I probably could implement any desired logic, but I honestly do not see that any logic would be better right now.
I honestly think this is fundamentally broken, unless you would e.g. say that the `dtype=np.ndarray` and then you stop as soon as you find the first level of nd-array.

**Comment by Gattocrucco** [comment]
Bookkeeping: bug still alive in numpy 2.2.3.

**Comment by seberg** [comment]
FWIW, I think this is a "won't fix".  What you want would be a plausible choice if starting from scratch.  But there is no choice without any caveats.  And I really think the current behavior is the most consistent we can get (at least that doesn't come with odd breakage).

The only way something could ever move here is with some new argument (_maybe_ even something like gh-5933/gh-19113, but not sure).
